AdoptAWidget: Shortcut

arps18 opened the pull request:
Added a DartPad sample to demonstrate the use of the `Shortcut` widget.

<!-- Specify what this pull request does -->
This pull request is:

- [x] A code sample
- [ ] More references
- [ ] More explanation

<!-- Add the AdoptAWidget issue that is assigned to you: -->
closes #69481

arps18 committed:
Added Interactive Sample

yash1200 commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -385,7 +385,45 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 /// when invoking an [Action] via a keyboard key combination that maps to an
 /// [Intent].
-///
+/// 
```
Remove extra spaces at the end of lines.

arps18 replied:
Thanks, @yash1200 😄

arps18 committed:
Removed Extra space

yash1200 commented:
Not just on one line, but extra spaces from all the lines which have one.

arps18 committed:
Removed Extra spaces

yash1200 commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+///     child: Actions(
+///       actions: <Type, Action<Intent>> {
+///         Increment: CallbackAction<Increment>(
+///           onInvoke: (Increment intent) => setState(()=> count = count + 1), 
```
Here also, same as above.(And also line 415)

arps18 committed:
Removed Extra Spaces

arps18 committed:
Removed Extra Space

AyushBherwani1998 commented on packages/flutter/lib/src/widgets/shortcuts.dart:
````diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 /// [Intent].
 ///
+/// {@tool dartpad --template=stateful_widget_material}
+/// ```dart import
````
```suggestion
/// ```dart imports
```

AyushBherwani1998 commented on packages/flutter/lib/src/widgets/shortcuts.dart:
````diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 ///
+/// {@tool dartpad --template=stateful_widget_material}
+/// ```dart import
+/// import package:flutter/services.dart
````
```suggestion
/// import 'package:flutter/services.dart';
```

arps18 committed:
Fixed Typo

arps18 committed:
Fixed Typos

arps18 committed:
Fixed Typo

obitodarky commented on packages/flutter/lib/src/widgets/shortcuts.dart:
````diff
@@ -388,7 +388,7 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 /// {@tool dartpad --template=stateful_widget_material}
 /// ```dart imports
-/// import package:flutter/services.dart
+/// import 'package:flutter/services.dart'
````
Looks like you forgot to add a `;` at the end.

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 /// when invoking an [Action] via a keyboard key combination that maps to an
 /// [Intent].
 ///
+/// {@tool dartpad --template=stateful_widget_material}
```
Let's use a scaffold center template:
```suggestion
/// {@tool dartpad --template=stateful_widget_scaffold_center}
```

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+///          onInvoke: (Decrement intent) => setState(()=> count = count - 1),
+///         ),
+///       },
+///       child: Center(
+///         child: Text('$count'),
+///       ),
```
You will need to wrap the text widget in a `Focus`, since the shortcuts only work when there's a keyboard-focusable widget. So, something like:
```dart
Focus(
  autofocus: true,
  child: Text('count: $count'),
),
```

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+/// Widget build(BuildContext context) {
+///   return Shortcuts(
+///     shortcuts: <LogicalKeySet, Intent> {
+///       LogicalKeySet(LogicalKeyboardKey.alt, LogicalKeyboardKey.keyI): Increment(),
+///       LogicalKeySet(LogicalKeyboardKey.alt, LogicalKeyboardKey.keyD): Decrement(),
```
Maybe use different key sets, because alt + d is already a shortcut for debugging on the web, so there's a conflict there. Maybe the + and - signs on the keyboard instead?
```suggestion
        LogicalKeySet(LogicalKeyboardKey.shift, LogicalKeyboardKey.equal): Increment(),
        LogicalKeySet(LogicalKeyboardKey.minus): Decrement(),
        LogicalKeySet(LogicalKeyboardKey.numpadAdd): Increment(),
        LogicalKeySet(LogicalKeyboardKey.numpadSubtract): Decrement(),
```

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+///     child: Actions(
+///       actions: <Type, Action<Intent>> {
+///         Increment: CallbackAction<Increment>(
+///          onInvoke: (Increment intent) => setState(()=> count = count + 1),
+///         ),
+///         Decrement: CallbackAction<Decrement>(
+///          onInvoke: (Decrement intent) => setState(()=> count = count - 1),
```
We should avoid using arrow functions because it assumes that the value is returned by the function passed into setState, when in reality nothing is.
```suggestion
///          onInvoke: (Increment intent) => setState(() { count = count + 1; }),
///         ),
///         Decrement: CallbackAction<Decrement>(
///          onInvoke: (Decrement intent) => setState(() { count = count - 1; }),
```

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+///          onInvoke: (Decrement intent) => setState(()=> count = count - 1),
+///         ),
+///       },
+///       child: Center(
```
We should provide some visual text here to recommend to users what they need to do to understand how this sample works. I'd recommend using a Row here that contains some text like:
```dart
Text('Add to the counter by pressing the keyboard numberpad "+" or Shift + "="')
Text('Decrement from the counter by pressing "-"')
```

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
````diff
@@ -386,6 +386,44 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 /// [Intent].
 ///
+/// {@tool dartpad --template=stateful_widget_material}
+/// ```dart imports
````
Right before this line, I think we should add some text to describe how to use this sample:

```dart
/// Here, we will use a [Shortcuts] and [Actions] widget to add and remove from a counter. 
/// This can be done by creating a child widget that is focused and pressing the logical key 
/// sets that have been defined in [Shortcuts] and defining the actions that each key set 
/// performs.
```

shihaohong reviewed (commented):
Left some comments on here. cc/ @gspencergoog who would know more about Shortcuts :)

arps18 committed:
Performed required changes

arps18 committed:
Performed required changes

arps18 committed:
Typo corrected

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,56 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
 /// when invoking an [Action] via a keyboard key combination that maps to an
 /// [Intent].
 ///
+/// Here, we will use a [Shortcuts] and [Actions] widget to add and remove from a counter.
```
Sorry, I should have been a little more specific. It should be in the dartpad :) Like this:
```dart
/// {@tool dartpad --template=stateful_widget_scaffold_center}
///
/// Here, we will use a [Shortcuts] and [Actions] widget to add and remove from a counter. 
/// This can be done by creating a child widget that is focused and pressing the logical key 
/// sets that have been defined in [Shortcuts] and defining the actions that each key set 
/// performs.
/// 
/// ```dart imports
/// ....
```

arps18 replied:
Sure, no issues! 😄

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,56 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+///     },
+///     child: Actions(
+///       actions: <Type, Action<Intent>> {
+///         Increment: CallbackAction<Increment>(
+///          onInvoke: (Increment intent) => setState(() { count = count + 1; }),
+///         ),
+///         Decrement: CallbackAction<Decrement>(
+///          onInvoke: (Decrement intent) => setState(() { count = count - 1; }),
+///         ),
```
```suggestion
///         Increment: CallbackAction<Increment>(
///           onInvoke: (Increment intent) => setState(() { count = count + 1; }),
///         ),
///         Decrement: CallbackAction<Decrement>(
///           onInvoke: (Decrement intent) => setState(() { count = count - 1; }),
///         ),
```

arps18 committed:
Edited Suggestions

shihaohong commented on packages/flutter/lib/src/widgets/shortcuts.dart:
```diff
@@ -386,6 +386,57 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
+/// Widget build(BuildContext context) {
+///   return Shortcuts(
+///     shortcuts: <LogicalKeySet, Intent> {
+///       LogicalKeySet(LogicalKeyboardKey.shift, LogicalKeyboardKey.numpadAdd): Increment(),
+///       LogicalKeySet(LogicalKeyboardKey.shift, LogicalKeyboardKey.numpadSubtract): Decrement(),
```
Just using shift + numpadAdd/Subtract is a little tough for some laptop users (for example, Macbooks do not have a numpad). Maybe include a set or two that is more accessible? Doesn't have to be + or -, even a letter on the keyboard will do.

arps18 replied:
How about `shift` and `keyK` for Increment and `shift` and `keyL` for Decrement?

shihaohong replied:
That should be good!

arps18 committed:
Edited Keys

arps18 committed:
Edited Typo

shihaohong reviewed (approved):
LGTM! Thanks for being so responsive with addressing code review feedback!

arps18 commented:
> LGTM! Thanks for being so responsive with addressing code review feedback!

Thank you so much for your valuable suggestions! I really enjoyed contributing 🎉 Looking for more contributions in the near future 😄

johnpryan commented:
@arps18 Thanks for participating in AdoptAWidget. We would like to send you a special thanks. If you are interested, please send an email to nyener at google dot com and provide a link to this issue.
